designs/resource_limits.rst

```diff
@@ -0,0 +1,179 @@
+---------------
+
+As a piece of security sensitive software, kanidm must be "available" (as defined by
+integrity, availability and confidentiality). This means that as a service we must
+be able to handle a large volume of potentially malicous traffic, and still able
+to serve legitimate requests without fault or failure.
+
+To achieve this, the resources of the server must be managed and distributed so allow
+potentially thousands of operations per second, while preventing exhaustion of those
+resources.
+
+Kanidm is structured as a database, where each request requires a process
+to resolve that query into an answer. This could be a request for authetication
+which is a true/false response, or a request for an identity so that we can
+determine their groups for authorisation, or even just a request to find
+someone's email address in a corporate directory context.
+
+Each operation requires resources to complete individually, but many operations
+can be processed in parallel.
+
+For a resource exhaustion to occur, this would be due to input from a client
+that consumes resources. This means the attack surface is any possible input
+and how the server interprets and processes that. In kanidm this could be
+a search filter for example - the query may be small, but it could be expensive
+to compute.
+
+CPU time
+========
+
+Every operation requires time on a CPU. A CPU logically is many threads which
+can have many operations in parallel, but also bound by a frequency and their
+performance as to how many instructions can be performed per second.
+
+This means that given a say 8 core CPU, each individual core may be able to
+complete 1000 operations per second, and combined this becomes 8x1000.
+
+If CPU time is exhausted, then operations are unable to be served. This means
+that we must manage the amount of time an operation is allowed to take
+on the system.
+
+Some types of queries can take more time than others. For example:
+
+* Unindexed searches which require full table scans.
+* Indexed searches that have a large number of results.
+* Write operations that affect many entries.
+
+These necesitate the following limits:
+
+* No unindexed searches allowed
+* Prevent searching on terms that do not exist (done, filter schema validation)
+* Limit on the number of entries that can be checked during a partial indexed search
+* Limit on the number of entries that can be loaded from a fully indexed search
+* Limit on the number of entries that can be written during a write operation
+* Reduce the size of allowed queries to prevent complex index access patterns
+
+Memory (RAM) Capacity
+=====================
+
+A system has finite ram amounts which must be shared between all CPU's of the system.
+Effective use of that RAM improves the speed of all operations, meaning that as a shared
+resource it can not be monopolised by a single operation.
+
+Each operation can be assumed to have access to RAM/N CPUS of ram. If an operation tempororily
```

Comment:
does N need to be a capital?

Reply:
I was going for ram divided by n[umber] of cpus, but maybe this is unclear?